lms/djangoapps/learner_recommendations/utils.py

```diff
@@ -224,3 +223,83 @@ def filter_recommended_courses(
             filtered_recommended_courses.append(course_data)
 
     return filtered_recommended_courses
+
+
+def get_programs_based_on_course(course_key, country_code, user):
+    """
+    Returns a program for the course. If a course is part of multiple programs,
+    this function returns the program with the highest price.
+    """
+    # Get the programs list for the course
+    programs = get_programs(course=course_key)
+
+    # check user is enrolled in different courses part of program
+    if programs:
+        enrollable_programs = []
+        for program in programs:
+            course_keys = []
+            for courses in program.get('courses'):
+                course_keys.append(courses.get('key'))
+            enrollable_course_keys = _remove_user_enrolled_course_keys(user, course_keys)
+            if enrollable_course_keys:
+                enrollable_programs.append(program)
+
+        non_restricted_programs = []
+        for program in enrollable_programs:
+            # Check for location restriction
+            if not _has_country_restrictions(program, country_code):
+                non_restricted_programs.append(program)
+
+        max_price_program = max(non_restricted_programs, key=lambda x: x['price_ranges'][0]['total'])
```

Comment:
- Repeated code with lots of loops
- Also you need to remove the program that a user is already enrolled in.

Sample of what it can look like (please test it):
```suggestion
    max_price = 0
    max_price_program = {}
    for program in programs:
        if program.get('status') != 'active' or _has_country_restrictions(program, country_code):
            continue

        # Skip the program a user is already enrolled in
        # <Add code here>

        price = program['price_ranges'][0]['total']
        if price > max_price:
            course_keys = [
                course['key']
                for course in programs.get('courses', [])
                if course.get('key') and course.get('key') != request_course
            ]
    
            if _remove_user_enrolled_course_keys(user, course_keys):
                max_price_program = program
                max_price = price

    if not max_price_program:
        return None
```